src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,186 @@
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> wanted = first; wanted != null; wanted = wanted.next) {
+                if (wanted.item == null) {
+                    unlink(wanted);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> wanted = first; wanted != null; wanted = wanted.next) {
+                if (object.equals(wanted.item)) {
+                    unlink(wanted);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder("[");
+        Node<T> current = first;
+        while (current != null) {
+            builder.append(current.item);
+            if (current.next != null) {
+                builder.append(", ");
+            } else {
+                builder.append("]");
+            }
+            current = current.next;
+        }
+        return builder.toString();
+    }
+
+    private void linkLast(T value) {
+        final Node<T> initLast = last;
+        final Node<T> newNode = new Node<>(initLast, value, null);
+        last = newNode;
+        if (initLast == null) {
+            first = newNode;
+        } else {
+            initLast.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> after) {
+        final Node<T> before = after.prev;
+        final Node<T> newNode = new Node<>(before, value, after);
+        after.prev = newNode;
+        if (before == null) {
+            first = newNode;
+        } else {
+            before.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> found) {
+        final T element = found.item;
+        final Node<T> next = found.next;
+        final Node<T> prev = found.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            found.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            found.next = null;
+        }
+        found.item = null;
+        size--;
+        return element;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Index is out of size range");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("There is no value on index: "
+                    + index);
+        }
+    }
+
+    private Node<T> findNode(int index) {
+        if (index < (size() / 2)) {
```

Comment:
Don't iterate your nodes from the head to the tail or vice versa. Optimize the iteration based on the index value to improve performance.